Userland/Libraries/LibWeb/CSS/Parser/DeprecatedCSSParser.cpp

```diff
@@ -229,6 +230,125 @@ static StringView parse_custom_property_name(const StringView& value)
     return value.substring_view(4, substring_length);
 }
 
+static StringView isolate_calc_expression(const StringView& value)
+{
+    if (!value.starts_with("calc(") || !value.ends_with(")"))
+        return {};
+    auto substring_length = value.length() - 5 - 1;
+    return value.substring_view(5, substring_length);
+}
+
+static CSS::CalculatedStyleValue::CalcSum parse_calc_expression(const StringView& expression_string)
+{
+    // First, tokenize
+
+    struct CalcToken {
+        enum class Type {
+            Undefined,
+            Number,
+            Unit,
+            Whitespace,
+            Plus,
+            Minus,
+            Asterisk,
+            Slash,
+            OpenBracket,
+            CloseBracket,
+        } type { Type::Undefined };
+        String value {};
+    };
+
+    size_t index = 0;
+    Vector<CalcToken> tokens;
+
+    char ch;
+    auto consume = [&index, &ch, &expression_string]() {
+        ch = expression_string[index];
+        index++;
+        return ch;
+    };
+
+    auto peek = [&index, &expression_string](int offset) {
+        auto new_index = index + offset - 1;
+        return expression_string[new_index];
+    };
+
+    while (index < expression_string.length()) {
+        consume();
+        // Number
+        if (((ch == '+' || ch == '-') && !isspace(peek(1))) || ch == '.' || isdigit(ch)) {
+            StringBuilder number;
+            while (ch == '+' || ch == '-' || ch == '.' || isdigit(ch)) {
+                number.append(ch);
+                consume();
+            }
+            index--;
+            tokens.append(CalcToken { CalcToken::Type::Number, number.to_string() });
```

Comment:
This style of "put token in vector and then iterate over vector" is extremely common in LibWeb's lexers
It's just a whole lot of perf issues for not a lot of gain (except maybe simplicity, but you're not storing any state in the control flow itself here)
I hope `calc(...)` expressions don't get long enough in practice to make this an issue.

Reply:
`calc()`s have at most 20 recursive `calc`s in them, so there are at most about 100 tokens, I don't think that's enough to become an issue. Also I think nobody uses that large `calc` expressions anyways.